power_grid_model_c/power_grid_model/include/power_grid_model/main_model_impl.hpp

```diff
@@ -528,23 +528,27 @@ class MainModelImpl<ExtraRetrievableTypes<ExtraRetrievableType...>, ComponentLis
         // cache component update order if possible
         bool const is_independent = MainModelImpl::is_update_independent(update_data);
 
-        return [&base_model, &exceptions, &infos, &calculation_fn, &result_data, &update_data,
+        std::shared_ptr<SequenceIdx const> all_scenarios_sequence{
```

Comment:
Why use a shared pointer? The thread does not need to own the resource right? We can just create an object at the main thread and capture the reference into the thread lambda. The main thread always out-lives the work threads.

Reply:
If we make it as a member variable of the `MainModel` instance in the main thread, then it looks weird because we reference a cache object that's not part of `this`.

On the other hand, if we define it locally in this function, then it will go out of lifetime the moment this function returns its lambda. The lambda then contains a reference to an out-of-lifetime object, which causes segmentation violations.

The only other potentially nice solution I see that does not use `shared_ptr` would have both the `all_scenarios_sequence` and the `cacheable_scenario_sequence` declared as member variables; the former in the main thread, and the latter in the worker thread

Reply:
Why will the local object go out of life before the work threads finish?

You first declare and create the local object `all_scenarios_sequence` locally in the function entry of `calculate`. And capture it by reference in the lambda's. The work threads will finish before the `calculate` finish.

Reply:
Ah I see the problem. This function is not the main entry point, but just a sub-routine to return the lambdas.

So my proposal:

1. Declare `all_scenarios_sequence` the entry point of `calculate`.
2. Pass `all_scenarios_sequence` by rvalue reference (which will be mutated) into this function `sub_batch_calculation_` returning the lambda.
3. Capture `all_scenarios_sequence` by const reference in the lambda.
4. Keep `all_scenarios_sequence` alive in the main entry point function until threads are finished.

Reply:
step 2 should be lvalue reference, otherwise ownership will be transferred to this function, after which it will go out of scope

Reply:
Yes, adjusted